### ISSUE TITLE
[Bug]: phpMyAdmin - Undefined variable $handle

### ISSUE THREAD

**Issue by xyz582** [beyondcode/herd-community#451]
### Operating system version

macOS Sonoma 14.3.1

### System architecture

ARM64 (M1, M2, etc)

### Herd Version

1.5.0 B:23

### PHP Version

PHP 8.3.3 

### Bug description

After installing phpmyadmin, this error pops up. Please help.

![Zrzut ekranu 2024-03-13 o 11 26 14](https://github.com/beyondcode/herd-community/assets/25344362/00f40fd9-3686-4d8f-ae60-abadc444af40)

```
Warning in ./../../../../../../../../Applications/Herd.app/Contents/Resources/valet/server.php#139
 Undefined variable $handle

Backtrace

{closure}()
```

Applications/Herd.app/Contents/Resources/valet/server.php code:
```
<?php

require_once './cli/includes/require-drivers.php';
require_once './cli/Valet/Server.php';

use Valet\Drivers\ValetDriver;
use Valet\Server;

/**
 * Define the user's "~/.config/valet" path.
 */
define('VALET_HOME_PATH', $_SERVER['HERD_HOME'].'/config/valet');
define('VALET_STATIC_PREFIX', '41c270e4-5535-4daa-b23e-c269744c2f45');

/**
 * Load the Valet configuration.
 */
$valetConfig = json_decode(
    file_get_contents(VALET_HOME_PATH.'/config.json'), true
);

/**
 * If the HTTP_HOST is an IP address, check the start of the REQUEST_URI for a
 * valid hostname, extract and use it as the effective HTTP_HOST in place
 * of the IP. It enables the use of Valet in a local network.
 */
if (Server::hostIsIpAddress($_SERVER['HTTP_HOST'])) {
    $uriForIpAddressExtraction = ltrim($_SERVER['REQUEST_URI'], '/');

    if ($host = Server::valetSiteFromIpAddressUri($uriForIpAddressExtraction, $valetConfig['tld'])) {
        $_SERVER['HTTP_HOST'] = $host;
        $_SERVER['REQUEST_URI'] = str_replace($host, '', $uriForIpAddressExtraction);
    }
}

$server = new Server($valetConfig);

/**
 * Parse the URI and site / host for the incoming request.
 */
$uri = Server::uriFromRequestUri($_SERVER['REQUEST_URI']);
$siteName = $server->siteNameFromHttpHost($_SERVER['HTTP_HOST']);
$valetSitePath = $server->sitePath($siteName);

if (is_null($valetSitePath) && is_null($valetSitePath = $server->defaultSitePath())) {
    Server::show404();
}

$valetSitePath = realpath($valetSitePath);

/**
 * Find the appropriate Valet driver for the request.
 */
$valetDriver = ValetDriver::assign($valetSitePath, $siteName, $uri);

if (! $valetDriver) {
    Server::show404();
}

/**
 * ngrok uses the X-Original-Host to store the forwarded hostname.
 */
if (isset($_SERVER['HTTP_X_ORIGINAL_HOST']) && ! isset($_SERVER['HTTP_X_FORWARDED_HOST'])) {
    $_SERVER['HTTP_X_FORWARDED_HOST'] = $_SERVER['HTTP_X_ORIGINAL_HOST'];
}

/**
 * Attempt to load server environment variables.
 */
$valetDriver->loadServerEnvironmentVariables(
    $valetSitePath, $siteName
);

/**
 * Allow driver to mutate incoming URL.
 */
$uri = $valetDriver->mutateUri($uri);

/**
 * Determine if the incoming request is for a static file.
 */
$isPhpFile = pathinfo($uri, PATHINFO_EXTENSION) === 'php';

if ($uri !== '/' && ! $isPhpFile && $staticFilePath = $valetDriver->isStaticFile($valetSitePath, $siteName, $uri)) {
    return $valetDriver->serveStaticFile($staticFilePath, $valetSitePath, $siteName, $uri);
}

/**
 * Allow for drivers to take pre-loading actions (e.g. setting server variables).
 */
$valetDriver->beforeLoading($valetSitePath, $siteName, $uri);

/**
 * Attempt to dispatch to a front controller.
 */
$frontControllerPath = $valetDriver->frontControllerPath(
    $valetSitePath, $siteName, $uri
);

if (! $frontControllerPath) {
    if (isset($valetConfig['directory-listing']) && $valetConfig['directory-listing'] == 'on') {
        Server::showDirectoryListing($valetSitePath, $uri);
    }

    Server::show404();
}

register_shutdown_function(function () use ($valetSitePath) {
    $configPath = $_SERVER['HERD_HOME'].'/config/';
    $herdConfigFile = $configPath.'/herd.json';

    if (file_exists(($herdConfigFile))) {
        $handle = fopen($herdConfigFile, 'r+');

        if (!flock($handle, LOCK_EX)) {
            fclose($handle);
            return;
        }


        try {
            $existingConfig = fread($handle, filesize($herdConfigFile));
            $existingConfig = json_decode($existingConfig, true);
            if (!is_array($existingConfig)) {
                $existingConfig = [];
            }
        } catch (\Throwable $e) {
            $existingConfig = [];
        }
    } else {
        $existingConfig = [];
    }

    try {
        $config = array_merge($existingConfig, [
            'lastSite' => $valetSitePath,
        ]);

        ftruncate($handle, 0);
        rewind($handle);
        fwrite($handle, json_encode($config));
        flock($handle, LOCK_UN);
        fclose($handle);
    } catch (\Throwable $e) {}
});

chdir(dirname($frontControllerPath));

$_SERVER['HERD_SITE_PATH'] = $valetSitePath;

require $frontControllerPath;
```

**Comment by mpociot** [comment]
This bug will be fixed in the next release.
You should be able to manually fix it by running `touch ~/Library/Application\ Support/Herd/config/herd.json`

**Comment by xyz582** [comment]
> This bug will be fixed in the next release. You should be able to manually fix it by running `touch ~/Library/Application\ Support/Herd/config/herd.json`

Thanks, creating the file fixes the problem. 

![Zrzut ekranu 2024-03-13 o 12 17 19](https://github.com/beyondcode/herd-community/assets/25344362/06d87a75-d8e0-4a06-967a-85eba2547048)

**Comment by n8finch** [comment]
> You should be able to manually fix it by running `touch ~/Library/Application\ Support/Herd/config/herd.json`

That works, thanks for the quick fix!

**Comment by coffeeyoghurt** [comment]
This did not solve my problem, it went from above error about undefined $handle to just HTTP 500 (no laravel error). Where can I find the error log?

**Comment by Odabasyusuf** [comment]
> This bug will be fixed in the next release. You should be able to manually fix it by running `touch ~/Library/Application\ Support/Herd/config/herd.json`

It worked for me. Thanks!

**Comment by nasrulkurniawan** [comment]
> touch ~/Library/Application\ Support/Herd/config/herd.json

thanks solved. works for me.

**Comment by ronzyfonzy** [comment]
I had to do create the `herd.json` in my user folder and then it solved the problen
```bash
touch /Users/$USER/Library/Application\ Support/Herd/config/herd.json
```

**Comment by themahdavi** [comment]
> This bug will be fixed in the next release. You should be able to manually fix it by running `touch ~/Library/Application\ Support/Herd/config/herd.json`

Thanks for support 💯

**Comment by mpociot** [comment]
This fixed with Herd 1.6.0

**Comment by musamaashraf** [comment]
> This bug will be fixed in the next release. You should be able to manually fix it by running `touch ~/Library/Application\ Support/Herd/config/herd.json`

Worked for me thanks!!
